### PR TITLE
Changelog: request params changed

### PR DESCRIPTION
**Pull request by blva**
[request param changed: pattern](https://github.com/Tufin/oasdiff/pull/343/commits/c3c8ac0b8c790b8de8d6684c6878d273ce496b5c)

[request param changed: min request param value increased/decreased](https://github.com/Tufin/oasdiff/pull/343/commits/97c2bd871f227987a4de454a66ed7f45856d04c5)

[request param changed: min items param value increased/decreased](https://github.com/Tufin/oasdiff/pull/343/commits/39361cd3a86bcd94e7112c2c22e3c0093b41a92f)

[request param changed: max request param value increased/decreased](https://github.com/Tufin/oasdiff/pull/343/commits/af2d70d4496f1f158ed851fca229da5ead1476c0)

[request param changed: max items request param value increased/decreased](https://github.com/Tufin/oasdiff/pull/343/commits/19b5bcb71b996bfea6d18394150077b3d6495bb8)

[request param changed: max length request param value increased/decreased](https://github.com/Tufin/oasdiff/pull/343/commits/4cd1c1d7355ec75aa4e500f00fc4148451e4803b) 

[request param changed: min length increased/decreased](https://github.com/Tufin/oasdiff/pull/343/commits/f22d77b51f8e9582381f76cc2f9d5f7c25758de8)

[request param changed: enum value added/removed](https://github.com/Tufin/oasdiff/pull/343/commits/739ea15af26aea5672f0756b7534eb44e4c44da6)

### CODE REVIEW

**Review by ciprian-tibulca — APPROVED** [review]

**Review by wtrocki — APPROVED** [review]

**Comment by blva** [comment]
@reuvenharrison cc for your review here since this is a big one and other pr is under discussion, thanks!

File: BREAKING-CHANGES-EXAMPLES.md [hunk 1]
```diff
@@ -140,6 +140,9 @@ These examples are automatically generated from unit tests.
 [renaming a path parameter is not breaking](checker/checker_breaking_test.go?plain=1#L135)  
 
 ## Examples of info-level changes for changelog
+[Adding pattern to request parameters](checker/check-request-parameter-pattern-added-or-changed_test.go?plain=1#L35)  
+[Changing pattern of request parameters](checker/check-request-parameter-pattern-added-or-changed_test.go?plain=1#L11)  
+[Removing pattern from request parameters](checker/check-request-parameter-pattern-added-or-changed_test.go?plain=1#L58)
```
Comment (reuvenharrison) [inline BREAKING-CHANGES-EXAMPLES.md]: Please change to lowercase

File: checker/check-request-parameter-enum-value-updated.go [hunk 2]
```diff
@@ -43,6 +43,17 @@ func RequestParameterEnumValueRemovedCheck(diffReport *diff.Diff, operationsSour
 							Source:      source,
 						})
 					}
+					for _, enumVal := range enumDiff.Added {
+						result = append(result, ApiChange{
+							Id:          "request-parameter-enum-value-added",
+							Level:       INFO,
+							Text:        fmt.Sprintf(config.i18n("request-parameter-enum-value-added"), enumVal, ColorizedValue(paramLocation), ColorizedValue(paramName)),
```
Comment (reuvenharrison) [inline checker/check-request-parameter-enum-value-updated.go]: I recommend to use ColorizedValue(enumVal), also in the other message, above

File: checker/check-request-parameter-pattern-added-or-changed_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/tufin/oasdiff/checker"
+	"github.com/tufin/oasdiff/diff"
+)
+
+// CL: Changing pattern of request parameters
```
Comment (reuvenharrison) [inline checker/check-request-parameter-pattern-added-or-changed_test.go]: lowercase

File: checker/check-request-parameter-pattern-added-or-changed_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		Comment:     "This is a warning because it is difficult to automatically analyze if the new pattern is a superset of the previous pattern (e.g. changed from '[0-9]+' to '[0-9]*')",
+		Level:       checker.WARN,
+		Operation:   "POST",
+		Path:        "/test",
+		Source:      "../data/checker/request_parameter_pattern_added_or_changed_base.yaml",
+		OperationId: "",
+	}, errs[0])
+}
+
+// CL: Adding pattern to request parameters
```
Comment (reuvenharrison) [inline checker/check-request-parameter-pattern-added-or-changed_test.go]: lowercase

File: checker/check-request-parameter-pattern-added-or-changed_test.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	require.NoError(t, err)
+
+	d, osm, err := diff.GetWithOperationsSourcesMap(getConfig(), s1, s2)
+	require.NoError(t, err)
+	errs := checker.CheckBackwardCompatibilityUntilLevel(singleCheckConfig(checker.RequestParameterPatternAddedOrChangedCheck), d, osm, checker.WARN)
+	require.Len(t, errs, 1)
+	require.Equal(t, checker.ApiChange{
+		Id:          "request-parameter-pattern-added",
+		Text:        "added the pattern '^\\w+$' to the 'query' request parameter 'category'",
+		Comment:     "This is a warning because it is difficult to automatically analyze if the new pattern is a superset of the previous pattern (e.g. changed from '[0-9]+' to '[0-9]*')",
```
Comment (reuvenharrison) [inline checker/check-request-parameter-pattern-added-or-changed_test.go:49]: This message makes sense for the use-case of changing the regex.
I would recommend to adjust the explanation for the specific use-case of adding a regex.

File: checker/check-request-parameter-pattern-added-or-changed_test.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		Level:       checker.WARN,
+		Operation:   "POST",
+		Path:        "/test",
+		Source:      "../data/checker/request_parameter_pattern_added_or_changed_base.yaml",
+		OperationId: "",
+	}, errs[0])
+}
+
+// CL: Removing pattern from request parameters
```
Comment (reuvenharrison) [inline checker/check-request-parameter-pattern-added-or-changed_test.go]: lowercase

File: checker/check-request-parameters-max-items-updated_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/tufin/oasdiff/checker"
+	"github.com/tufin/oasdiff/diff"
+)
+
+func TestRequestParameterMaxItemsIncreased(t *testing.T) {
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-items-updated_test.go:12]: add a comment so it gets documented

File: checker/check-request-parameters-max-items-updated_test.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		Comment:     "",
+		Level:       checker.INFO,
+		Operation:   "POST",
+		Path:        "/api/v1.0/groups",
+		Source:      "../data/checker/request_parameter_max_items_updated_revision.yaml",
+		OperationId: "createOneGroup",
+	}, errs[0])
+}
+
+func TestRequestParameterMaxItemsDecreased(t *testing.T) {
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-items-updated_test.go:35]: add a comment so it gets documented

File: checker/check-request-parameters-max-items-updated.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		for operation, operationItem := range pathItem.OperationsDiff.Modified {
+			if operationItem.ParametersDiff == nil {
+				continue
+			}
+			for paramLocation, paramDiffs := range operationItem.ParametersDiff.Modified {
+				for paramName, paramDiff := range paramDiffs {
+					if paramDiff.SchemaDiff == nil {
+						continue
+					}
+					maxItemsDiff := paramDiff.SchemaDiff.ItemsDiff.MaxItemsDiff
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-items-updated.go:27]: check if paramDiff.SchemaDiff.ItemsDiff == nil before using it

File: checker/check-request-parameters-max-items-updated.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+					if maxItemsDiff == nil {
+						continue
+					}
+					if maxItemsDiff.From == nil ||
+						maxItemsDiff.To == nil {
+						continue
+					}
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-items-updated.go:34]: I'm not sure but it's possible that From==nil means no max
In that case, changing nil to a number would, in effect, be breaking

File: checker/check-request-parameters-max-length-updated_test.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/tufin/oasdiff/checker"
+	"github.com/tufin/oasdiff/diff"
+)
+
+func TestRequestParameterMaxLengthIncreasedCheck(t *testing.T) {
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-length-updated_test.go:12]: add comment to document the use-case

File: checker/check-request-parameters-max-length-updated_test.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+		Text:        "for the 'query' request parameter 'category', the maxLength was increased from '10' to '15'",
+		Level:       checker.INFO,
+		Operation:   "POST",
+		Path:        "/test",
+		Source:      "../data/checker/request_parameter_max_length_updated_revision.yaml",
+		OperationId: "",
+	}, errs[0])
+}
+
+func TestRequestParameterMaxLengthDecreasedCheck(t *testing.T) {
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-length-updated_test.go:34]: add comment to document the use-case

File: checker/check-request-parameters-max-length-updated_test.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	s2, err := open("../data/checker/request_parameter_max_length_updated_revision.yaml")
+	require.NoError(t, err)
+
+	d, osm, err := diff.GetWithOperationsSourcesMap(getConfig(), s1, s2)
+	require.NoError(t, err)
+	errs := checker.CheckBackwardCompatibilityUntilLevel(singleCheckConfig(checker.RequestParameterMaxLengthUpdatedCheck), d, osm, checker.INFO)
+	require.Len(t, errs, 1)
+	require.Equal(t, checker.ApiChange{
+		Id:          "request-parameter-max-length-increased",
+		Text:        "for the 'query' request parameter 'category', the maxLength was increased from '10' to '15'",
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-length-updated_test.go:24]: I recommend to change to:
"for the 'query' request parameter 'category', maxLength was increased from '10' to '15'",

(without "the")

File: checker/check-request-parameters-max-length-updated_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	s2, err := open("../data/checker/request_parameter_max_length_updated_base.yaml")
+	require.NoError(t, err)
+
+	d, osm, err := diff.GetWithOperationsSourcesMap(getConfig(), s1, s2)
+	require.NoError(t, err)
+	errs := checker.CheckBackwardCompatibilityUntilLevel(singleCheckConfig(checker.RequestParameterMaxLengthUpdatedCheck), d, osm, checker.INFO)
+	require.Len(t, errs, 1)
+	require.Equal(t, checker.ApiChange{
+		Id:          "request-parameter-max-length-decreased",
+		Text:        "for the 'query' request parameter 'category', the maxLength was decreased from '15' to '10'",
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-length-updated_test.go:46]: same, no need for "the" before maxLength

File: checker/check-request-parameters-max-updated_test.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+	s2, err := open("../data/checker/request_parameter_max_updated_revision.yaml")
+	require.NoError(t, err)
+
+	d, osm, err := diff.GetWithOperationsSourcesMap(getConfig(), s1, s2)
+	require.NoError(t, err)
+	errs := checker.CheckBackwardCompatibilityUntilLevel(singleCheckConfig(checker.RequestParameterMaxUpdatedCheck), d, osm, checker.INFO)
+	require.Len(t, errs, 1)
+	require.Equal(t, checker.ApiChange{
+		Id:          "request-parameter-max-increased",
+		Text:        "for the 'query' request parameter 'category', the max was increased from '5.00' to '10.00'",
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-updated_test.go:24]: no need for "the" before max

File: checker/check-request-parameters-max-updated_test.go [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+	s2, err := open("../data/checker/request_parameter_max_updated_base.yaml")
+	require.NoError(t, err)
+
+	d, osm, err := diff.GetWithOperationsSourcesMap(getConfig(), s1, s2)
+	require.NoError(t, err)
+	errs := checker.CheckBackwardCompatibilityUntilLevel(singleCheckConfig(checker.RequestParameterMaxUpdatedCheck), d, osm, checker.INFO)
+	require.Len(t, errs, 1)
+	require.Equal(t, checker.ApiChange{
+		Id:          "request-parameter-max-decreased",
+		Text:        "for the 'query' request parameter 'category', the max was decreased from '10.00' to '5.00'",
```
Comment (reuvenharrison) [inline checker/check-request-parameters-max-updated_test.go:46]: again, no need for "the" before max

File: checker/check-request-parameters-min-items-updated_test.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+	s2, err := open("../data/checker/request_parameter_min_items_increased_revision.yaml")
+	require.NoError(t, err)
+
+	d, osm, err := diff.GetWithOperationsSourcesMap(getConfig(), s1, s2)
+	require.NoError(t, err)
+	errs := checker.CheckBackwardCompatibilityUntilLevel(singleCheckConfig(checker.RequestParameterMinItemsUpdatedCheck), d, osm, checker.ERR)
+	require.Len(t, errs, 1)
+	require.Equal(t, checker.ApiChange{
+		Id:          "request-parameter-min-items-increased",
+		Text:        "for the 'query' request parameter 'category', the minItems was increased from '2' to '3'",
```
Comment (reuvenharrison) [inline checker/check-request-parameters-min-items-updated_test.go:24]: the minItems -> minItems

File: checker/check-request-parameters-min-items-updated_test.go [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+	s2, err := open("../data/checker/request_parameter_min_items_increased_base.yaml")
+	require.NoError(t, err)
+
+	d, osm, err := diff.GetWithOperationsSourcesMap(getConfig(), s1, s2)
+	require.NoError(t, err)
+	errs := checker.CheckBackwardCompatibilityUntilLevel(singleCheckConfig(checker.RequestParameterMinItemsUpdatedCheck), d, osm, checker.INFO)
+	require.Len(t, errs, 1)
+	require.Equal(t, checker.ApiChange{
+		Id:          "request-parameter-min-items-decreased",
+		Text:        "for the 'query' request parameter 'category', the minItems was decreased from '3' to '2'",
```
Comment (reuvenharrison) [inline checker/check-request-parameters-min-items-updated_test.go:46]: the minItems -> minItems

**Review by reuvenharrison — CHANGES_REQUESTED** [review]
Please see the comments inline

File: checker/check-request-parameter-enum-value-updated.go (hunk 2 above)
Comment (blva) [reply checker/check-request-parameter-enum-value-updated.go]: Fixing for `request-parameter-enum-value-added` should be fine as we are introducing it now. 

But updating the other one, which already existed may mean we'll start adding '' on the enum value message, which is a change of output and could cause regressions to users if they have ignore-err with those values, right? What is the best option in that case? We can probably do the change but make sure we release a new minor version after this.

File: checker/check-request-parameters-max-items-updated.go (hunk 10 above)
Comment (blva) [reply checker/check-request-parameters-max-items-updated.go:34]: I think we have max-items set checker for this, here i just followed same pattern as other max-item checkers

File: checker/check-request-parameters-max-length-updated_test.go (hunk 13 above)
Comment (blva) [reply checker/check-request-parameters-max-length-updated_test.go:24]: do we really want to change it when other checkers follow the same pattern? 

e.g. `request-parameter-max-length-decreased: for the %s request parameter %s, the maxLength was decreased from %s to %s`, I've no strong opinion but usually prefer to follow previous patterns or change them all, but changing them means potential issues on `err-ignore` functionality.

File: checker/check-request-parameter-enum-value-updated.go (hunk 2 above)
Comment (reuvenharrison) [reply checker/check-request-parameter-enum-value-updated.go]: Yes, it may break ignore files. Good point.
I would still recommend to make this correction, at the worst case, people will need to reconfigure their ignore files.
We will add a note about this in the release.

Regarding the version number:
Normally, we would need to increase the major version to signify incompatible API changes. But we can't do that because changing a major version in golang is too difficult.
Also note, that we have two kinds of clients: cmd-line clients and go package clients, and changes that may break one type of client may not break the other one.
To summarize, we are incapable of doing proper version management until we separate the go package from the cmd-line tool.

**Review by reuvenharrison — APPROVED** [review]
